### PR TITLE
Disable test generation for synthetic methods of data classes #1191

### LINKED ISSUE #1191: We shouldn't generate tests for synthetic methods of data classes
**Description**

Kotlin automatically generates some methods for data classes, however, we shouldn't test them unless they are overridden manually.

**To Reproduce**

Create data class with no methods and launch action on it:

```Kotlin
data class Datalass(val x: Int, val y: Int)
```

**Expected behavior**

Action is disabled

**Actual behavior**

Dialog appears with synthetic methods being proposed for testing

**Visual proofs (screenshots, logs, images)**

**Additional context**

The issue looks like a "Kotlin edition" of #457

### PR DESCRIPTION
**Pull request by volivan239**
# Description

Added filters so that autogenerated by Kotlin methods of data classes are not tested by plugin/cli/contestEstimator

Fixes #1191

## Type of Change

- Minor bug fix (non-breaking small changes)

# How Has This Been Tested? 

## Manual Scenario 

Tested the behavior in plugin -- autogenerated methods are not added to memberlist (unless they are overridden).
In cli tests for autogenerated methods are also not generated.

# Checklist (remove irrelevant options):

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [x] No new warnings
- [ ] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
     if (method.classId.isEnum)
-        method.humanReadableName.substringBefore('(') in KnownSyntheticMethodNames.enumSyntheticMethodNames
-    else
-        false
+        return method.name in KnownSyntheticMethodNames.enumSyntheticMethodNames
+
+    if (method.classId.isData)
+        return KnownSyntheticMethodNames.dataClassSyntheticMethodNames.any { it.matches(method.name) }
+
+    return false
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt:14]: `when` may be better here, I suppose

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt [hunk 2]
```diff
@@ -17,4 +21,14 @@ fun isKnownSyntheticMethod(method: ExecutableId): Boolean =
...
     /** List with names of enum methods that are autogenerated */
     val enumSyntheticMethodNames = listOf("values", "valueOf")
+
+    /** List with names of data classes methods that are autogenerated in Kotlin */
+    val dataClassSyntheticMethodNames = listOf(
+        "equals",
+        "hashCode",
+        "toString",
+        "copy",
+        "copy\\\$default",
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt:31]: What is this method? Add some comments about it, please

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt [hunk 3]
```diff
@@ -17,4 +21,14 @@ fun isKnownSyntheticMethod(method: ExecutableId): Boolean =
 private object KnownSyntheticMethodNames {
     /** List with names of enum methods that are autogenerated */
     val enumSyntheticMethodNames = listOf("values", "valueOf")
+
+    /** List with names of data classes methods that are autogenerated in Kotlin */
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt:25]: Mention that they are actually regexps, please

**Review by Damtev — CHANGES_REQUESTED** [review]
Add a bit more docs, please

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt [hunk 4]
```diff
@@ -1,14 +1,18 @@
...
-fun isKnownSyntheticMethod(method: ExecutableId): Boolean =
+fun isKnownSyntheticMethod(method: ExecutableId): Boolean {
     if (method.classId.isEnum)
-        method.humanReadableName.substringBefore('(') in KnownSyntheticMethodNames.enumSyntheticMethodNames
-    else
-        false
+        return method.name in KnownSyntheticMethodNames.enumSyntheticMethodNames
+
+    if (method.classId.isData)
+        return KnownSyntheticMethodNames.dataClassSyntheticMethodNames.any { it.matches(method.name) }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt:12]: Am I right that CLI will not generate test methods for overridden "synthetic" methods in data classes (`equals`, for example)?

**Review by Damtev — CHANGES_REQUESTED** [review]
What about overridden methods in data classes for CLI?

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt (hunk 1 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: Is it possible here? `isEnum` and `isData` are just properties...

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt (hunk 2 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: It is JVM-synthetic and needed when copy is called with default parameters, after last commit it is not needed here.

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt (hunk 3 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: Fixed, also changed the property name for more clarity

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt (hunk 4 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: Yes, and it seems difficult (or impossible) to check if the method is not compiler-generated but overridden by user (but if you know the way, I would appreciate hearing it)

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt (hunk 1 above)
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: Just a simple one, like
```kotlin
when {
    method.classId.isEnum -> ...
    method.classId.isData -> ...
    else -> ...
}
```

Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SyntheticMethods.kt]: Done

**Review by Damtev — APPROVED** [review]

**Pull request merged**